java-checks/src/main/java/org/sonar/java/checks/BasicAuthCheck.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2019 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.Arrays;
+import java.util.List;
+import org.sonar.check.Rule;
+import org.sonar.java.checks.methods.AbstractMethodDetection;
+import org.sonar.java.matcher.MethodMatcher;
+import org.sonar.plugins.java.api.tree.BinaryExpressionTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import static org.sonar.java.checks.helpers.ConstantUtils.resolveAsStringConstant;
+
+@Rule(key = "S2647")
+public class BasicAuthCheck extends AbstractMethodDetection {
+
+  @Override
+  protected List<MethodMatcher> getMethodInvocationMatchers() {
+    return Arrays.asList(
+      MethodMatcher.create().typeDefinition("org.apache.http.message.AbstractHttpMessage").name("setHeader").withAnyParameters(),
+      MethodMatcher.create().typeDefinition("java.net.URLConnection").name("setRequestProperty").withAnyParameters()
+      );
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree mit) {
+    if ("Authorization".equals(resolveAsStringConstant(mit.arguments().get(0)))) {
+      ExpressionTree arg = mostLeft(mit.arguments().get(1));
+      String authentication = resolveAsStringConstant(arg);
+      if (authentication != null && authentication.startsWith("Basic")) {
+        reportIssue(arg, "Use a more secure method than basic authentication.");
+      }
+    }
+  }
+
+  private static ExpressionTree mostLeft(ExpressionTree arg) {
+    ExpressionTree res = arg;
+    while (res.is(Tree.Kind.PLUS)) {
```

Comment:
What about skipping parenthesis, so that we still raise issue for the following:
`conn.setRequestProperty("Authorization", ("Basic " + encoding + encoding))`